File: A_Circumference_of_a_Tree.cpp
#include "stdc++.h"

using namespace std;

class Node {
public:
    vector<Node> adj;
    int dist;
};

void dfs (Node from, vector<Node> nodes) {
    for (auto nn : nodes) {
        nn.dist = -1;
    }
    from.dist = 0;
    deque<Node> s;
    s.push_back(from);
    while (!s.empty()) {
        Node next = s.front();
        s.pop_front();
        for (auto nn : next.adj) {
            if (nn.dist == -1) {
                nn.dist = next.dist + 1;
                s.push_back(nn);
            }
        }
    }
}

Node farthest (vector<Node> nodes) {
    Node max = nodes[0];
    for (auto nn : nodes) {
        if (nn.dist > max.dist) {
            max = nn;
        }
    }
    return max;
}

int main () {
    int n; cin >> n;
    vector<Node> nodes(n);
    for (int i = 1; i < n; i++) {
        int a, b; cin >> a >> b;
        a--;
        b--;
        nodes[a].adj.push_back(nodes[b]);
        nodes[b].adj.push_back(nodes[a]);
    }
    dfs(nodes[0], nodes);
    Node far = farthest(nodes);
    dfs(far, nodes);
    cout << 3 * farthest(nodes).dist << endl;
}

File: A_Circumference_of_a_Tree.java
import java.io.*;
import java.util.*;

public class A_Circumference_of_a_Tree {

    static class Node {
        ArrayList<Node> adj = new ArrayList<Node>();
        int dist;
    }

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        if (n == 1) {
            System.out.println(0);
            return;
        }
        Node[] nodes = new Node[n];
        for (int i = 0; i < n; i++) nodes[i] = new Node();
        for (int i = 1; i < n; i++) {
           int a = sc.nextInt() - 1, b = sc.nextInt() - 1;
           nodes[a].adj.add(nodes[b]);
           nodes[b].adj.add(nodes[a]);
        }
        bfs(nodes[0], nodes);
        Node farthest = farthest(nodes);
        bfs(farthest, nodes);
        System.out.println(3 * farthest(nodes).dist);
        sc.close();
    }

    static Node farthest (Node[] nodes) {
        Node max = nodes[0];
        for (Node nn : nodes) {
            if (nn.dist > max.dist) {
                max = nn;
            }
        }
        return max;
    }

    static void bfs (Node from, Node[] nodes) {
        for (Node nn : nodes) nn.dist = -1;
        from.dist = 0;
        ArrayDeque<Node> q = new ArrayDeque<Node>();
        q.add(from);
        while (!q.isEmpty()) {
            Node next = q.remove();
            for (Node nn : next.adj) {
                if (nn.dist == -1) {
                    nn.dist = next.dist + 1;
                    q.add(nn);
                }
            }
        }
    }
}